docs: add how-to for cross compiling autotools (#5350)
canonical/snapcraft: docs/howto/craft-a-snap/example-cross-compiled-app.rst

.. _example-cross-compiled-app:

Example cross-compiled app
==========================

Projects built using the :ref:`craft_parts_autotools_plugin` can be cross-compiled,
allowing software to be developed on one architecture and built for a number of target
architectures.

Example project file for example-curl
-------------------------------------

The following code comprises the project file for a project that uses libcurl. The
complete code can be found in `this repository
<https://github.com/snapcraft-docs/hello-curl>`_.

.. collapse:: hello-curl project

  .. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
      :caption: snapcraft.yaml
      :language: yaml

Enable cross-compilation
------------------------

> The :ref:`platforms <how-to-architectures>` key is used to specify which architectures
> should be used when building a snap. The ``platforms`` key can also be used to specify
> sets of architectures using the ``build-on`` and ``build-for`` keys. In the simplest
> case, this can be used to cross-compile on one architecture for a single, different
> architecture:

Reviewer: ```suggestion
The :literalref:`platforms <how-to-architectures>` key specifies which architectures
should be used when building a snap. The ``platforms`` key can also specify
sets of architectures with the ``build-on`` and ``build-for`` keys. In the simplest
case, this can be used to cross-compile on one architecture for a single, different
architecture:
```
Just some more concise language :)

.. code-block:: yaml
    :caption: snapcraft.yaml

    platforms:
      - build-on: amd64
        build-for: armhf

In this example, Snapcraft will only build the project on the AMD64 architecture,
and it will only build it for the armhf architecture.

> The ``hello-curl`` project uses the ``platforms`` key to list many cross-compilation
> targets:

Reviewer: ```suggestion
In the ``hello-curl`` project, the ``platforms`` key lists several cross-compilation
targets:
```

----------------------------------

When building for a particular architecture, Snapcraft will initialise the
``CRAFT_ARCH_TRIPLET_BUILD_FOR`` environment variable in the build environment. This variable
describes the platform and architecture that Autotools uses to configure
cross-compilation. For more information on environment variables, see `environment
variables <https://snapcraft.io/docs/parts-environment-variables>`_.

A project building with the Autotools plugin can adjust the options passed to the
``configure`` script for a given part with the plugin's
``autotools-configure-parameters`` key.

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: snapcraft.yaml of hello-curl
   :language: yaml
   :start-at: autotools-configure-parameters:

Satisfy the dependencies
------------------------

The ``build-packages`` key specifies which packages will supply the toolchain and
libraries necessary for cross-compilation.

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: snapcraft.yaml of hello-curl
   :language: yaml
   :start-at: build-packages:
   :end-before: stage-packages:

> `Package repositories <https://snapcraft.io/docs/package-repositories>`_ can also be
> specified for the target platform. The hello-curl project defines additional
> repositories for target platforms using the ``package-repositories`` key and the
> ``CRAFT_ARCH_BUILD_FOR`` environment variables:

Reviewer: ```suggestion
`Package repositories <https://snapcraft.io/docs/package-repositories>`_ can also be
specified for the target platform with the ``package-repositories`` key and the
``CRAFT_ARCH_BUILD_FOR`` environment variable.
```

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: snapcraft.yaml of hello-curl
   :language: yaml
   :start-at: package-repositories:
   :end-at: url:

To select the correct library for the target platform, include these packages in the
``stage-packages`` key with the ``package:repo`` syntax.

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: snapcraft.yaml of hello-curl
   :language: yaml
   :start-at: stage-packages:
   :end-before: autotools-configure-parameters:

The ``build-packages`` and ``stage-packages`` keys additionally support the `advanced
grammar <https://snapcraft.io/docs/snapcraft-advanced-grammar>`_ keys, which allow
further customisation of packages installed per-platform.

Build the snap
--------------

> The snap is built in the usual way, by running ``snapcraft`` in the project directory:

Reviewer: ```suggestion
To build the snap, navigate to the project directory and run:
```

Gotta sneak one last nickpick in there :) ignore if you like

.. _example-cross-compiled-app:

Example Cross-Compiled App
==========================

Projects built using the :ref:`craft_parts_autotools_plugin` can be cross-compiled,
allowing software to be developed on one architecture and built for a number of target
architectures.

Overview
--------

There are a number of steps to enable cross-compilation for an autotools project:

#. Specify the architectures used.
#. Adjust the options passed to the ``configure`` script.
#. Include packages to support cross-compilation.
#. Handle linker warnings.

Example project file for example-curl
-------------------------------------

The following code comprises the project file for a project that uses ``libcurl``. The

> complete code can be found in `this repository <https://github.com/mr-cal/hello-curl>`_.

Author: Placeholder URL. I'd like to get this in https://github.com/snapcraft-docs, trying to find the right person to talk to for that.

.. _example-cross-compiled-app:

> Example Cross-Compiled App

Reviewer 2: ```suggestion
Example cross-compiled app
```

==========================

Projects built using the :ref:`craft_parts_autotools_plugin` can be cross-compiled,
allowing software to be developed on one architecture and built for a number of target
architectures.

Overview

> --------

Reviewer 2: I think the document would be stronger if we removed this section.

> There are a number of steps to enable cross-compilation for an autotools project:

Reviewer 2: ```suggestion
There are a number of steps to enable cross-compilation for an Autotools project:
```

All instances should be "Autotools".

#. Specify the architectures used.
#. Adjust the options passed to the ``configure`` script.
#. Include packages to support cross-compilation.

> #. Handle linker warnings.

Reviewer 2: ```suggestion
#. Handle linter warnings.
```

Example project file for example-curl
-------------------------------------

> The following code comprises the project file for a project that uses ``libcurl``. The

Reviewer 2: ```suggestion
The following code comprises the project file for a project that uses libcurl. The
```

Some commands, programs, and library names have become genericised as verbs, tools, or products. For example, sudo started as a command but has since entered common parlance. Similarly for curl and libcurl, the word might look like a command, but it's actually a term (product name) in its own right, as its [home page](https://curl.se/docs/manpage.html) demonstrates.

I'll publish this elsewhere later, but the rule goes something like:

- If we're describing the command and how it's used, or the command isn't well-known enough to be recognised as a regular word, surround the word with code fences. For example, "When you run `curl -i`..."
- If we're discussing a project, product, a common verb that is also a command, or the general idea behind the word, write the word plainly. For example, "Download and install curl."

What qualifies as "common" isn't always clear, so it will be up to us to decide.

complete code can be found in `this repository <https://github.com/mr-cal/hello-curl>`_.

.. collapse:: hello-curl project

  .. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
      :caption: snapcraft.yaml
      :language: yaml

The ``platforms`` key

> -------------------------

Reviewer 2: I think this heading would be better if it spoke to a task and contained an imperative verb. How about "Target an architecture"?

The :ref:`platforms <architectures-howto>` key is normally used to specify which
architectures should be used when building a snap. For example, in the following case
the build architecture is the same as the target architecture:

.. code:: yaml

    platforms:
      - amd64

It is also possible to specify different architectures for the build process and the
target system. This enables cross-compilation to be managed by Snapcraft.

Enable cross-compilation
------------------------

The ``platforms`` key can also be used to specify sets of architectures using the

> ``build-on`` and ``build-for`` sets. In the simplest case, this can be used to

Reviewer 2: ```suggestion
``build-on`` and ``build-for`` keys. In the simplest case, this can be used to
```

cross-compile on one architecture for a single, different architecture:

.. code:: yaml

    platforms:
      - build-on: amd64

>         build-for: armhf

Author: I'm getting an error that the `:caption:` option doesn't exist for code directives. Is there a different directive I should be using?

>         build-for: armhf

Author: Aha - apparently that option only exists for `code-block` directives, which it seems I should've been using anyways.

>         build-for: armhf

Reviewer: My mistake! I forgot to change that in the suggestion

>         build-for: armhf

Reviewer 2: I should have caught that, too. We should sweep that whole pattern out of the code.

In this example, Snapcraft will only build the project on the ``amd64`` architecture,

> and it will only build it for the ``armhf`` architecture.

Reviewer 2: In line with my earlier comment about curl, `amd64` and `armhf` are machine names for the semi-official terms "AMD64" and "armhf".

A corollary would be the following fabricated sentences:

> To build for the `win` and `darwin` platforms, list `win` and `darwin` in the `platforms` key.

Compared to:

> To build for Windows and macOS, list `win` and `darwin` in the `platforms` key.

The ``hello-curl`` project uses the ``platforms`` key to list many cross-compilation
targets:

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml

>    :caption: platforms

Reviewer 2: In keeping with other docs, what do you think about making this `snapcraft.yaml of hello-curl`?

   :language: yaml
   :start-at: platforms:
   :end-before: lint:

Adjust the autotools configuration
----------------------------------

When building for a particular architecture, Snapcraft will initialise the
``CRAFT_ARCH_TRIPLET_BUILD_FOR`` environment variable in the build environment. This variable
describes the platform and architecture that autotools uses to configure
cross-compilation. For more information on environment variables, see `environment
variables <https://snapcraft.io/docs/parts-environment-variables>`_.

> A project that uses the autotools plugin can adjust the options passed to the
> ``configure`` script for a given part by using the ``autotools-configure-parameters``
> key that the plugin provides. In the example project, the ``hello-curl`` part is
> configured this way:

Reviewer: ```suggestion
A project building with the autotools plugin can adjust the options passed to the
``configure`` script for a given part with the plugin's
``autotools-configure-parameters`` key.
```

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: autotools-configure-parameters
   :language: yaml
   :start-at: autotools-configure-parameters:

> Supply the dependencies

Reviewer 2: > Supply

What do you think about "satisfy" instead? "Supply" makes me think we're doing the work of copying the packages manually.

-----------------------

It is necessary to specify which packages will supply the toolchain and libraries
needed for cross-compilation. In the example project, the ``hello-curl`` part also
includes these definitions:

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: build-packages
   :language: yaml
   :start-at: build-packages:
   :end-before: stage-packages:

`Package repositories <https://snapcraft.io/docs/package-repositories>`_ can also be

> specified for the target platform. The example project defines additional repositories

Reviewer 2: ```suggestion
specified for the target platform. The hello-curl project defines additional repositories
```

for target platforms using the ``package-repositories`` key and the

> ``CRAFT_ARCH_BUILD_FOR`` environment variables:

Reviewer: Is there a way to do this other than how it is done in the example project?

Author: No, not any official "non-hacky" ways, at least.

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: package-repositories
   :language: yaml
   :start-at: package-repositories:
   :end-at: url:

> These repositories can then be used with the ``package:repo`` syntax to select the
> correct library for the target platform. The example project uses this in the
> ``stage-packages`` key:

Reviewer: ```suggestion
To select the correct library for the target platform, include these packages in the
``stage-packages`` key with the ``package:repo`` syntax.
```
Much like the `build-attributes` comment, disregard if the usage of the `stage-packages` key is specific to this example.

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml
   :caption: stage-packages
   :language: yaml
   :start-at: stage-packages:
   :end-before: autotools-configure-parameters:

The ``build-packages`` and ``stage-packages`` keys additionally support the `advanced
grammar <https://snapcraft.io/docs/snapcraft-advanced-grammar>`_ keys, which allow
further customisation of packages installed per-platform.

Build the snap
--------------

> The snap is built in the usual way, by running ``snapcraft`` in the project directory:
> 
> .. terminal::
>    :input: snapcraft

Reviewer: ```suggestion
To build the snap, navigate to the project directory and run:

.. code-block:: yaml

    snapcraft
```
Since we aren't showing any output, I don't think the terminal directive is necessary.

Reviewer 2: Agreed, let's use `code-block` here.

Because linters are enabled by default for ``core24`` snaps, this will produce warnings
like the following:

> .. code::
> 
>     not a dynamic executable
>     arm-binfmt-P: Could not open '/lib/ld-linux-armhf.so.3': No such file or directory
>     Unable to determine library dependencies for '/root/prime/usr/bin/autotools-cross-compile-example'

Reviewer: ```suggestion
.. terminal::
    :input: snapcraft
    
    not a dynamic executable
    arm-binfmt-P: Could not open '/lib/ld-linux-armhf.so.3': No such file or directory
    Unable to determine library dependencies for '/root/prime/usr/bin/autotools-cross-compile-example'
```
I think this is the better place for the terminal to show the error messages in context.

This is because snapcraft is unable to resolve the dependencies for the target
architecture using the library linter.

Since the :file:`/lib/ld-linux-armhf.so.3` library will be present in the base of the
target system, these warnings can be suppressed for this case by including a ``lint``
section in the project file:

.. literalinclude:: ../code/craft-a-snap/example-cross-compiled-app-recipe.yaml

>    :caption: linters

Reviewer: ```suggestion
   :caption: snapcraft.yaml
```

   :language: yaml
   :start-at: lint:
   :end-at: - library

Rebuilding the snap should result in a build process without warnings and a snap in the
project directory called :file:`hello-curl_1.0_armhf.snap` or similar. The ``_armhf``

> component of the file name indicates that the snap has been built for that architecture.

Reviewer 2: The only substantive thing I'd change in this doc is this last line. There are stronger ways to impart this idea and further thinking on a user. Consider:

> Rebuild the snap, and this time it should finish without warnings. A `.snap` file for each target architecture is deposited at the root of the project directory. Each snap is distinguished by the architecture name that's included in its file name.

As an addendum to the above, shouldn't the build also generate `*_arm64.snap` and `*_riscv.snap`?

Author: Yes, it does include those other files. I was trying to single out one example to prevent it being too long, but now that you point it out, I think that could've been confusing. I've listed all three in a condensed way now instead.
